\documentclass{article}


\usepackage{arxiv}

\usepackage[utf8]{inputenc} % allow utf-8 input
\usepackage[T1]{fontenc}    % use 8-bit T1 fonts
\usepackage{hyperref}       % hyperlinks
\usepackage{url}            % simple URL typesetting
\usepackage{booktabs}       % professional-quality tables
\usepackage{amsfonts}       % blackboard math symbols
\usepackage{nicefrac}       % compact symbols for 1/2, etc.
\usepackage{microtype}      % microtypography
\usepackage{lipsum}
\usepackage{graphicx}

\title{The Sankoff Algorithm for Phylogeographics}


\author{
  Julia Fischer \\
  Universität Tübingen \\
  \texttt{juli.fischer@student.uni-tuebingen.de} \\
  6039174 \\
  \And
  Peter Heringer \\
  Universität Tübingen \\
  \texttt{peter.heringer@student.uni-tuebingen.de} \\
  6109174 \\
  \And 
  Michael Mederer \\
  Universität Tübingen \\
  \texttt{michael.mederer@student.uni-tuebingen.de} \\
  XXXXXXX \\
  \And 
  Felix Seidel \\
  Universität Tübingen \\
  \texttt{felix.seidel@student.uni-tuebingen.de} \\
  5969276 \\
}

\begin{document}
\maketitle

\begin{abstract}
\end{abstract}


\section{Introduction}
Given a rooted tree $T$ with labeled leaves, the small parsimony problem is about
finding labels for the internal nodes of the tree such that the changes from an
internal nodes to all of its children are minimal
\cite{jonesIntroductionBioinformaticsAlgorithms2004}.
Among others, the Sankoff algorithm can be used to solve the small parsimony
problem \cite{sankoffMinimalMutationTrees1975}. 

\textbf{TODO: maybe short description of how Sankoff works?}

The Sankoff algorithm is usually applied the broader scope of
inferring phylogenetic trees using cost matrices that model the transition
between different DNA or RNA nucleotides
\cite{jonesIntroductionBioinformaticsAlgorithms2004}. By using different cost
matrices, the algorithm can also be used to model other biological questions,
for example in the Camin-Sokal-Parsimony \cite{caminMethodDeducingBranching1965}
or in the Dollo-Parsimony \cite{farrisPhylogeneticAnalysisDollo2022}. More
recently, the Sankoff algorithm was also used to infer the geographic origins of
the 2009 H1N1 pandemic using a distance matrix that represents the distances of
various international airports
\cite{reimeringPhylogeographicReconstructionUsing2020}. In this, the authors use
a phylogenetic tree that represents the relationship between different strains
of the virus and map each taxon to the airport that is closest to the sampling
location. Then, three different cost matrices are utilized:

\begin{enumerate}
  \item Equal distance: Airports have a distance of 1 to each other
  \item Geographic distance: The actual geographic distances between different
  airports is used.
  \item Effective distances: A infection is more likely to spread from
  airport $A$ to airport $B$ if there are many people traveling from $A$ to
  $B$. Thus, the effective distances are calculated by using passenger data
  for each combination of airports.
\end{enumerate}

Using those cost matrices for airports and for the corresponding countries, the
labels for the internal nodes of the tree are calculated and thus their
geographic position.

In the following report, we implement the method outlined in
\cite{reimeringPhylogeographicReconstructionUsing2020} and recreate the findings
by using the original data.

\section{Results}



\subsection{Fr\'{e}chet distances}
\begin{itemize}
    \item Fr\'{e}chet tree distances used to quantify geographic differences between spread paths
    \item We compare our reconstructed trees with the ones from the paper
    \item by claculating discrete Fr\'{e}chet tree distances using geographic distances between locations
    \item method compares the paths of locations from the root of each leaf node, calculates discrete Fr\'{e}chet distances between them and corrects the distance for each node by the number of paths
    \item we use Rscript from the authors for calculation \cite{reimeringFrechetTreeDistance2018}
\end{itemize}

Our inferred trees versus the trees from the paper:
\begin{itemize}
    \item geographic distances: 123.8497
    \item effective distances: 0
\end{itemize}

The Fr\'{e}chet distance between our trees:
\begin{itemize}
    \item effective airport vs geographic airport: 65149
    \item effective country vs geographic country: 12307.03
    \item airport to country comparison not possible due to distance matrix
\end{itemize}

\section{Material and Methods}

\section{Discussion}


\bibliographystyle{unsrt}  
\bibliography{references}  %%% Remove comment to use the external .bib file (using bibtex).
%%% and comment out the ``thebibliography'' section.

\end{document}

